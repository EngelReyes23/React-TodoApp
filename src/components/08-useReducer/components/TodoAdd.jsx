import React from "react";
import PropTypes from "prop-types";
import { useForm } from "../../../Hooks/useForm";

const TodoAdd = ({ handleAddTodo }) => {
	// Custom Hook
	const [{ description }, handleInputChange, reset] = useForm({
		description: "",
	});

	const handleFormSubmit = (e) => {
		e.preventDefault();

		// Verifica que el valor no este vacio
		if (description.trim().length <= 2) return;

		const newtodo = {
			id: new Date().getTime(),
			description,
			done: false,
		};
		handleAddTodo(newtodo);
		reset();
	};

	return (
		<>
			<h4>Add TODO</h4>
			<hr />
			<form action="" onSubmit={handleFormSubmit}>
				<input
					type="text"
					name="description"
					placeholder="Aprender..."
					autoComplete="off"
					className="form-control"
					onChange={handleInputChange}
					value={description}
				/>
				<div className="d-grid gap-2">
					<button type="submit" className="btn btn-outline-primary mt-1">
						Add
					</button>
				</div>
			</form>
		</>
	);
};

TodoAdd.propTypes = {
	handleAddTodo: PropTypes.func.isRequired,
};

export default TodoAdd;
