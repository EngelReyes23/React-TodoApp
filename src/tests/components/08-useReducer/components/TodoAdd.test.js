import { shallow } from "enzyme";
import TodoAdd from "../../../../components/08-useReducer/components/TodoAdd";

describe("Pruebas en <TodoAdd />", () => {
	// preparacion Test
	const handleAddTodo = jest.fn();
	const wrapper = shallow(<TodoAdd handleAddTodo={handleAddTodo} />);

	// 1er Test - Renderizar
	test("Debe de mostrarse correctamente", () => {
		expect(wrapper).toMatchSnapshot(); // snapshot
	});

	//2do Test - no debe de llamar a handleAddTodo
	test("No debe de llamar a handleAddTodo", () => {
		wrapper.find("button").simulate("click");
		expect(handleAddTodo).toHaveBeenCalledTimes(0);
	});

	//3er Test - Debe de llamar a handleAddTodo
	test("Debe de llamar a handleAddTodo", () => {
		// const wrapper = mount(<TodoAdd handleAddTodo={handleAddTodo} />);
		wrapper.find("input").simulate("change", {
			target: {
				name: "description",
				value: "Hola Mundo",
			},
		});
		wrapper.find("form").simulate("submit", { preventDefault() {} });
		expect(handleAddTodo).toHaveBeenCalledTimes(1);

		// verificnado que se llamara con un argumento esperado
		expect(handleAddTodo).toHaveBeenCalledWith({
			id: expect.any(Number),
			description: "Hola Mundo",
			done: false,
		});

		// cuando se reinicia el formulario
		expect(wrapper.find("input").prop("value")).toBe("");
	});
});
