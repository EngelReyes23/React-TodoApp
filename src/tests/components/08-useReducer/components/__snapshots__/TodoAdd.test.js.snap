// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd /> Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
    Add TODO
  </h4>
  <hr />
  <form
    action=""
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender..."
      type="text"
      value=""
    />
    <div
      className="d-grid gap-2"
    >
      <button
        className="btn btn-outline-primary mt-1"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</Fragment>
`;
